### ISSUE TITLE
docker-compose exec -d (detached) doesn't return control of terminal

### ISSUE THREAD

**Issue by rickshawhobo** [docker/compose#4690]
According to the documentation `docker-compose exec -d` should run in detached mode, but after running with -d the control of the terminal is not returned. Am I missing something?

**Comment by shin-** [comment]
It works for me.

You're missing the version of Compose you're using, the command you're running, and basically any other details that would allow me to attempt reproducing your issue.

**Comment by rickshawhobo** [comment]
```
docker-compose --version
docker-compose version 1.11.2, build dfed245
docker --version
Docker version 17.03.0-ce, build 60ccb22
```

The command is basically any.

`docker-compose -d exec [service_name] bash` hangs and doesn't return the terminal.
`docker-compose -d exec [service_name] ls` hangs and doesn't return the terminal.

**Comment by giper45** [comment]
Maybe a similar issue : let's suppose the following script "telnet_client.sh"  into a  container named "telnet-client"  : 
```
while true; do
        if [ $# -eq 4 ]
        then
        TELNET_SERVER=$1
        NAME=$2
        PASSWORD=$3
        COMMANDS=$4
        eval '{ echo;
                sleep 3;
                echo $NAME ;
                sleep 1;
                echo $PASSWORD ;
                sleep 1;
                echo $COMMANDS ;
                sleep 1; }' | telnet $TELNET_SERVER


        else
                echo "Not enough arguments ! "
        fi

done;
```


If I call the following : 
``` docker-compose exec -d telnet_client bash -c '/telnet_client.sh telnet_server jack jack ls' ```


The control in the terminal is not returned.  
Otherwise , if I call this : 
``` docker exec -d lab3_telnet_client_1 bash -c '/telnet_client.sh telnet_server jack jack ls' ```  
the script is called correctly in background mode .

**Comment by giper45** [comment]
A possible workaround could be 
``` 
docker exec -d $(docker-compose ps -q service_name) command
```

**Comment by shin-** [comment]
@giper45 
Does it work with `docker-compose exec -T -d telnet_client bash -c '/telnet_client.sh telnet_server jack jack ls'`?

By default, Compose allocates a TTY, while the Docker CLI doesn't, so the two commands you're using are not equivalent unless you add `-T` in the Compose case. This might be the reason for the discrepancies you're seeing.

**Comment by giper45** [comment]
@shin-  
Unfortunately it doesn't work , I just try ... this is the definition of container in docker-compose yaml : 
```
 telnet_client:
    image: '...debian_bot:latest'
    stdin_open: true
    tty: true
    networks:
      internal_network:
        ipv4_address: 192.168.1.3
    cap_add:
      - NET_ADMIN
```   
Maybe stdin_open and tty options could cause this ? Anyway, with docker exec command I resolved,   let me know if I can give you further infos in order to  discover if there is a problem !

**Comment by shin-** [comment]
Thanks, I was able to reproduce and identify the issue. This should be fixed by #4737
